File: Baekjoon/Class3/01.11723 집합.cpp
#include <iostream>
using namespace std;

class BitmaskSet { // unsigned long long 비트마스크 세트
private:
    unsigned long long _set;
public:
    BitmaskSet() : _set(0) {};
    void add(int data) {
        _set |= 1ULL << (data - 1);
    }
    void remove(int data) {
        _set &= ~(1ULL << (data - 1));
    }
    bool check(int data) const {
        return (_set & (1ULL << (data - 1))) != 0;
    }
    void toggle(int data) {
        _set ^= 1ULL << (data - 1);
    }
    void all() {
        _set = ~0ULL;
    }
    void empty() {
        _set = 0;
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int m;
    BitmaskSet s;

    cin >> m;
    while (m--) {
        int num;
        string command;
        cin >> command;
        if (command == "add") {
            cin >> num;
            s.add(num);
        }
        else if (command == "remove") {
            cin >> num;
            s.remove(num);
        }
        else if (command == "check") {
            cin >> num;
            cout << s.check(num) << '\n';
        }
        else if (command == "toggle") {
            cin >> num;
            s.toggle(num);
        }
        else if (command == "all") {
            s.all();
        }
        else if (command == "empty") {
            s.empty();
        }
    }
    return 0;
}

/* 라이브러리 사용
#include <iostream>
#include <bitset>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int m;
    bitset<21> s;

    cin >> m;
    while (m--) {
        int num;
        string command;
        cin >> command;
        if (command == "add") {
            cin >> num;
            s.set(num);
        }
        else if (command == "remove") {
            cin >> num;
            s.reset(num);
        }
        else if (command == "check") {
            cin >> num;
            cout << s.test(num) << '\n';
        }
        else if (command == "toggle") {
            cin >> num;
            s.flip(num);
        }
        else if (command == "all") {
            s.set();
        }
        else if (command == "empty") {
            s.reset();
        }
    }
    return 0;
}
*/

File: Baekjoon/Class3/01.11723 집합.java
import java.io.*;
import java.util.StringTokenizer;

class Main {
    private static class BitmaskSet { // long 비트마스크 세트
        private long set;
        public BitmaskSet() { set = 0; }
        public void add(int data) {
            set |= 1L << (data - 1);
        }
        public void remove(int data) {
            set &= ~(1L << (data - 1));
        }
        public boolean check(int data) {
            return (set & (1L << (data - 1))) != 0;
        }
        public void toggle(int data) {
            set ^= 1L << (data - 1);
        }
        public void all() {
            set = ~0L;
        }
        public void empty() {
            set = 0;
        }
    }

    public static void main(String[] args) throws IOException {
        int m;
        BitmaskSet s = new BitmaskSet();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        m = Integer.parseInt(br.readLine());
        for(int i = 0; i < m; i++) {
            int num;
            st = new StringTokenizer(br.readLine());
            String command = st.nextToken();
            switch (command) {
                case "add":
                    num = Integer.parseInt(st.nextToken());
                    s.add(num);
                    break;
                case "remove":
                    num = Integer.parseInt(st.nextToken());
                    s.remove(num);
                    break;
                case "check":
                    num = Integer.parseInt(st.nextToken());
                    sb.append(s.check(num) ? 1 : 0).append('\n');
                    break;
                case "toggle":
                    num = Integer.parseInt(st.nextToken());
                    s.toggle(num);
                    break;
                case "all":
                    s.all();
                    break;
                case "empty":
                    s.empty();
                    break;
            }
        }
        System.out.print(sb);
        br.close();
    }
}

/* 라이브러리 사용
import java.io.*;
import java.util.BitSet;
import java.util.StringTokenizer;

class Main {
    public static void main(String[] args) throws IOException {
        int m;
        BitSet s = new BitSet(21);
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        m = Integer.parseInt(br.readLine());
        for(int i = 0; i < m; i++) {
            int num;
            st = new StringTokenizer(br.readLine());
            String command = st.nextToken();
            switch (command) {
                case "add":
                    num = Integer.parseInt(st.nextToken());
                    s.set(num);
                    break;
                case "remove":
                    num = Integer.parseInt(st.nextToken());
                    s.clear(num);
                    break;
                case "check":
                    num = Integer.parseInt(st.nextToken());
                    sb.append(s.get(num) ? 1 : 0).append('\n');
                    break;
                case "toggle":
                    num = Integer.parseInt(st.nextToken());
                    s.flip(num);
                    break;
                case "all":
                    s.set(1, 21);
                    break;
                case "empty":
                    s.clear();
                    break;
            }
        }
        System.out.print(sb);
        br.close();
    }
}
*/